### ISSUE TITLE
select from ? where XYZ

### ISSUE THREAD

**Issue by Usnul** [orientechnologies/orientdb#7787]
## Expected behavior  
given string "Metadata" -> select from Metadata where XYZ 

## Actual behavior  
select from METADATA where XYZ

## Steps to reproduce  
final OCommandSQL iCommand = new OCommandSQL("SELECT FROM ? WHERE `uuid`= ? LIMIT 1")
final OCommandRequest request = rawGraph.command(iCommand);
final Iterable<ODocument> result = request.execute(vertexClassName, "00000000-0000-000c-0000-000000000016");

here's a dump:
```
com.orientechnologies.orient.core.exception.OQueryParsingException: Error on parsing query at position #11: Error on parsing query
Query:  "Metadata" WHERE `uuid` = "00000000-0000-000c-0000-000000000016" LIMIT 1
----------------^
	DB name="test"

	at com.orientechnologies.orient.core.sql.filter.OSQLTarget.<init>(OSQLTarget.java:74)
	at com.orientechnologies.orient.core.sql.OSQLEngine.parseTarget(OSQLEngine.java:462)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:281)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:96)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:58)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:39)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:2559)
	at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:69)
[omitted part of the stack]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: com.orientechnologies.orient.core.exception.OCommandExecutionException: Class '"METADATA"' was not found in database 'test'
	DB name="test"
	at com.orientechnologies.orient.core.sql.filter.OSQLTarget.extractTargets(OSQLTarget.java:273)
	at com.orientechnologies.orient.core.sql.filter.OSQLTarget.<init>(OSQLTarget.java:64)
	... 41 more
```
maven:
```
<dependencies>
        <!-- START: OrientDB dependencies -->
        <dependency>
            <groupId>com.tinkerpop.gremlin</groupId>
            <artifactId>gremlin-java</artifactId>
            <version>2.6.0</version>
        </dependency>
        <dependency>
            <groupId>com.tinkerpop.blueprints</groupId>
            <artifactId>blueprints-core</artifactId>
            <version>2.6.0</version>
        </dependency>
        <dependency>
            <groupId>com.orientechnologies</groupId>
            <artifactId>orientdb-client</artifactId>
            <version>2.2.17</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <!-- This gives us object binding -->
            <groupId>com.orientechnologies</groupId>
            <artifactId>orientdb-object</artifactId>
            <version>2.2.17</version>
        </dependency>
        <dependency>
            <groupId>com.orientechnologies</groupId>
            <artifactId>orientdb-graphdb</artifactId>
            <version>2.2.17</version>
        </dependency>
        <dependency>
            <groupId>com.orientechnologies</groupId>
            <artifactId>orientdb-core</artifactId>
            <version>2.2.18</version>
            <scope>compile</scope>
        </dependency>
        <!-- END: OrientDB dependencies -->
</dependencies>
```
I am an architect on a project which relies on a graph database for its features. I have chosen OrientDB after reading your documentation. We are currently in the process of purchasing support from you.

My experience with Orient so far: utter garbage.

**Comment by lvca** [comment]
Dear @Usnul, first of all, saying that your experience is "utter garbage" is rude and offensive, especially because:
- you indicated OrientDB 2.3, but it doesn't exist this version
- in your pom.xml you mixed different version of OrientDB, 2.2.17 and 2.2.18. Why?
- what are you trying to accomplish? It's not clear
- if you're trying to query the OrientDB metadata, then your syntax is wrong. Please read the documentation: https://orientdb.com/docs/2.2/SQL.html#query-metadata
- If you want to query from a class with name "Metadata", then you have to create it first

**Comment by Usnul** [comment]
Hey @ivka, thank you for a quick response. I am not surprised you find those words offensive based on your reputation, but that is beside the issue. 

I have tried 2.2.17 core also, with the same effect. I do not fully comprehend what you mean when you say that 2.3 doesn't exist. But you are the expert, not I. I have attempted to provide as much information as possible, I'll be happy to add anything that  might help.

To address your assumptions: there exists a class named "Metadata" which extends class "V" in the database. I am trying to retrieve a record (vertex) of that class. As I have multiple classes which I would like to query using custom field "uuid" which all of these classes posess - I would like to use a prepared query with class name being parameterized. This is a very frequently used query in our use cases, so it seemed like a good candidate for optimization.

Main issue is that class name being supplied in parameters is "Metadata", but class being checked for existence by the parser seems to be "METADATA", which doesn't exist.

To come back to my review of experience with your product, i can elaborate on points beyond this failure if you wish. Unless you only welcome positive feedback, in which case I apologize for my earlier comment.

**Comment by lvca** [comment]
@Usnul this is an issue tracker, so 99% of the thousands of issues can't be positive, right? Rarely somebody opens an issue to report that everything is fine. So it's not about positive feedback, it's about the "utter garbage". This is an Open Source technology. If you think it's that bad, you can use something else or you can improve it.

The latest version is 2.2.28, but if you can find OrientDB v2.3, please send me the link. Maybe there is another team that is working in parallel with us :-)

Coming back to the issue, in v.2.2.x the prepared statements don't accept a variable in place of the target. I think v3.0 already support this, but in v2.2 you have to concatenate the class name as a string, sorry.

**Comment by Usnul** [comment]
@ivka thank you for clarification.

There are quite a few things that I expected to be true in orient based on documentation that have proven otherwise in past several months working with orient. Namely one of the biggest reasons to choose orient for me was support of gremlin. In practice gremlin queries in orient are thousands of times slower, server-side gremlin interpreter is slower than just using tonkerpop pipes in java on client side, in the end the performance is just not viable for large graphs and interactive use cases. Using gremlin graph interface in orient leads to violations of "happens before" principle where subsequent reads after writes (with commit) are not visible sometimes (sporadically). That's to name a few issues. Orient by itself with orient flavour of sql is quite competent and i like some of the internal architecture, it is elegant and consistent.

**Comment by laa** [comment]
Hi, @Usnul , I am quite interested in issue "Using gremlin graph interface in orient leads to violations of "happens before" principle where subsequent reads after writes (with commit) are not visible sometimes (sporadically)." could you provide any code which can reproduce it, at least sporadically?

About gremlin queries, that is a problem of all databases which support gremlin (I have read a statistic of different benchmarks for databases which support gremlin). By its nature, it is very hard to optimise gremlin queries. Since 3.0 version we bring the support of graph elements and queries of course into the core but with the usage of native ODB queries. In general, if you intend to use gremlin to achieve vendor independence you have to be ready to pay a price (that is obvious of course) but this price will be high for any database because as I mentioned above it is very hard to optimise gremlin queries. So if you have a mood to establish relations with us which are based on the mutual respect we are eager to check all issues which you have and help you. But alas gremlin queries will be always much slower I am sorry does not matter which database you will take. That is quotation from research paper (the independent research of course)

> For Neo4j, the Gremlin interface introduces up to two orders of magnitude of performance degradation compared to the native Cypher interface. Similarly, Postgres (SQL) significantly outperforms Sqlg (Gremlin) even though both systems have the same underlying data model and storage engine.

**Comment by Usnul** [comment]
I thought so originally too, that gremlin structures do not map well onto traditional indices with classes (tables) and keys. After having thought about it some more - I believe queries can be made faster using alternative indexing schemes, using inferred classes and having reverse lookups from field name to class. As far as "g.V" idiom goes - there are two things that can be done : 
- pushdown predicates (take predicates further down in the pipeline and use them to restrict element set.
- introduce additional interfaces for V beyond V(id), you already do that with getVerticesOfClass (i believe that's what the interface is called)

there are 2 problems that need to be solved:
- prepare a collection of elements for pipe (efficiently)
- process elements through the pipe (efficiently)

the first I believe is a larger problem in terms of performance impact, because blindly iterating through everything is just not what a database is sought for.

Ultimately it is about the focus, if you want to have a great performance with gremlin - it is achievable, but it would require additional abstractions in the database engine on a fairly low level and if such abstractions would be costly to implement and/or interfere with performance of existing interfaces (Orient SQL etc.) you may not wish to consider them.

I have done some benchmarks with Orient using OSQL, Pipes and Gremlin on server side (interpreter plugin, enabled via orient-settings.xml I believe, "Supported Languages"?) with results such that a certain query was taking about 0.2ms with OSQL, 10s with Pipes and about 30 seconds with gremlin on server side, it is clear to me that there are plenty of areas ripe for optimization based on that.

I am currently traveling and don't have access to any of the relevant code, I might re-engage you on performance topics once I return if there is still interest from your side.

As far as respect goes - I do respect authors of the codebase of Orient; like I said - I find a lot of what I have read to be very much to my liking, as far as code goes. When it comes to the product however, I find marketing material of Orient to be objectionable in its claims - which is okay, since I typically do not rely on such materials. If you read through Orient's documentation, however, it pains a picture of certain features being production ready and in my opinion (everyone has their own I guess) those features are not working well enough to be called anything but experimental when it comes to a database product.
In short - i have respect for developers of the product, but the product management and marketing has fostered mainly distrust. I hope this clarifies my position; I felt it was necessary to clarify it since it has been commented on several times.

**Comment by laa** [comment]
@luigidellaquila @maggiolo00 could you check the previous comment? There are interesting thoughts there IMHO.

**Comment by luigidellaquila** [comment]
Hi @Usnul 

Thank you very much for this analysis. For sure on Gremlin side there is a lot of room for optimization, we started supporting Apache TinkerPop 3 just a few months ago, so I hope we'll be able to make it better in the near future.

Predicate pushdown is more or less what we do in very basic cases in TinkerPop 2, actually we re-write part of the Gremlin query as an SQL statement, so that it can be preprocessed on the server side, with an important performance gain. The problem here is that Gremlin and SQL have a very different paradigm, so translating one query in the other one is not trivial task.
In V 3.0 (RC in very few weeks hopefully) we completely re-wrote our query execution planning, now it's more flexible and easier to optimize and extend; probably one of next steps will be to leverage this new architecture to optimize Gremlin too. I cannot give you and ETA though, we have other things in high priority on our pipeline...

Thanks

Luigi
